crawl-ref/source/transform.cc

```diff
@@ -1974,15 +1974,13 @@ bool transform(int pow, transformation_type which_trans, bool involuntary,
     if (crawl_state.which_god_acting() == GOD_XOM)
        you.transform_uncancellable = true;
 
-    // Re-check terrain now that be may no longer be swimming or flying.
-    if (was_flying && !you.airborne()
-                   || feat_is_water(grd(you.pos()))
-                      && (which_trans == TRAN_BLADE_HANDS
-                          || which_trans == TRAN_APPENDAGE)
-                      && you.species == SP_MERFOLK)
-    {
+    // Land the player if we stopped flying.
+    if (was_flying && !you.airborne())
         move_player_to_grid(you.pos(), false);
-    }
+
+    // Update merfolk swimming for the form change.
+    if (you.species == SP_MERFOLK)
+        merfolk_check_swimming(false);
```

Comment:
this will be redundant if we also called `move_player_to_grid` for landing, but as all the `merfolk_*_swimming` methods are idempotent, I think being more clear about what's needed outweighs the performance cost of calling this twice?

(same for `untransform` below)